include/ginkgo/core/log/performance_hint.hpp

```diff
@@ -0,0 +1,133 @@
+ *
+ * @ingroup log
+ */
+class PerformanceHint : public Logger {
+public:
+    void on_allocation_completed(const Executor* exec,
+                                 const size_type& num_bytes,
+                                 const uintptr& location) const override;
+
+    void on_free_completed(const Executor* exec,
+                           const uintptr& location) const override;
+
+    void on_copy_completed(const Executor* from, const Executor* to,
+                           const uintptr& location_from,
+                           const uintptr& location_to,
+                           const size_type& num_bytes) const override;
+
+    /**
+     * Writes out the cross-executor writes and allocations that have been
+     * stored so far.
+     */
+    void print_status() const;
+
+    /**
+     * Creates a PerformanceHint logger. This dynamically allocates the memory,
+     * constructs the object and returns an std::unique_ptr to this object.
+     *
+     * @param exec  the executor
+     * @param os  the stream used for this logger
+     * @param allocation_size_limit  ignore allocations below this limit (bytes)
+     * @param copy_size_limit  ignore copies below this limit (bytes)
+     * @param histogram_max_size  how many allocation sizes and/or pointers to
+     *                            keep track of at most?
+     *
+     * @return an std::unique_ptr to the the constructed object
+     */
+    static std::unique_ptr<PerformanceHint> create(
+        std::shared_ptr<const Executor> exec, std::ostream& os = std::cerr,
+        size_type allocation_size_limit = 16, size_type copy_size_limit = 16,
+        size_type histogram_max_size = 1024)
+    {
+        return std::unique_ptr<PerformanceHint>(
+            new PerformanceHint(exec, os, allocation_size_limit,
+                                copy_size_limit, histogram_max_size));
+    }
+
+protected:
+    explicit PerformanceHint(std::shared_ptr<const gko::Executor> exec,
+                             std::ostream& os, size_type allocation_size_limit,
+                             size_type copy_size_limit,
+                             size_type histogram_max_size)
+        : Logger(exec, mask_),
+          os_(&os),
+          allocation_size_limit_{allocation_size_limit},
+          copy_size_limit_{copy_size_limit},
+          histogram_max_size{histogram_max_size}
+    {}
+
+private:
+    // set a breakpoint here if you want to see where the output comes from!
+    std::ostream& log() const;
+
+    std::ostream* os_;
+    mutable std::unordered_map<uintptr_t, size_type> allocation_sizes_;
+    mutable std::unordered_map<size_type, int> allocation_histogram_;
+    mutable std::unordered_map<uintptr_t, int> copy_src_histogram_;
+    mutable std::unordered_map<uintptr_t, int> copy_dst_histogram_;
+    size_type allocation_size_limit_ = 16;
+    size_type copy_size_limit_ = 16;
+    size_type histogram_max_size = 1024;
```

Comment:
Probably the defaults are not necessary since you already specify them in the create function.